src/overview/actions.js

```diff
@@ -257,12 +269,83 @@ export const showTags = index => (dispatch, getState) => {
 }
 
 export const filterTag = tag => (dispatch, getState) => {
-    const query = selectors.query(getState())
+    const state = getState()
+    const query = selectors.query(state)
+    const filterTags = selectors.filterTags(state)
+
     const transformedTag = `#${tag.split(' ').join('+')} `
 
-    const newQuery = query.includes(transformedTag)
+    const newQuery = query.includes(transformedTag.slice(0, -1))
         ? query.replace(transformedTag, '') // Either remove it, if already there
         : transformedTag + query // or prepend it, if not there
 
-    dispatch(setQuery(newQuery))
+    dispatch(setQueryTagsDomains(newQuery))
+}
+
+export const setQueryTagsDomains = (input, isEnter) => (dispatch, getState) => {
+    const state = getState()
+    const filterTags = selectors.filterTags(state)
+    const filterDomains = selectors.filterDomains(state)
+    const filterStatus = selectors.showFilter(state)
+
+    const domains = []
+    const tags = []
+
+    const terms = input.toLowerCase().match(/\S+/g) || []
+
+    terms.reduce((terms, currTerm) => {
+        if (constants.DOMAIN_TLD_PATTERN.test(currTerm)) {
+            domains.push(currTerm)
+        }
+
+        if (constants.HASH_TAG_PATTERN.test(currTerm)) {
+            // Slice off '#' prefix and replace any '+' with space char
+            currTerm = currTerm
+                .slice(1)
+                .split('+')
+                .join(' ')
+            tags.push(currTerm)
+        }
+    }, [])
+    /* when the input's last character is space or when we enter the query just add or remove the tag/domain
+       from the query and filters and if there is any # char with tags or if there is any domain
+    */
+    if (
+        (input[input.length - 1] === ' ' || isEnter) &&
+        ((input.includes('#') && tags.length) || domains.length)
```

Comment:
This part is __really__ complicated and seems to be having quite an impact on UI perf when you are typing. All this really needs to do:
1. run `String.prototype.match()` on the query state to extract patterns
2. if `null`, don't do anything, else iterate over results:
  2.1. dispatch add/del actions to update relevant filter state
3. dispatch query updating action, removing any step 2 results from the query state

You would run it twice for tags and domains, with the pattern used in step 1 differing.
This main condition can probably just be reduced to the `(input[input.length - 1] === ' ' || isEnter)` part (upon further understanding, that `reduce` part earlier doesn't need to be there at all, so no need to check the lengths).

Is there ever a time when filter state can be mutated but the filters are hidden? If not, either set the filters to show in the reducers for add/del tag/domain actions or wrap the actions in a thunk (so `addTagFilter` always dispatched with `showFilter` for example - reducer way probably simpler).

It may also be wise to add a `toggleTagFilter` action where the reducer decides to add or remove based on prev state, to reduce all the decision making here, although just a suggestion